### PR TITLE
Validate parse trees in dialogue manager + modules path fix

### PR DESCRIPTION
**Pull request by Rebecca-Qian**
# Description

Add a hook in dialogue manager to validate parse trees from ground truth or semantic parser.

If the parse tree fails the grammar spec, treat as NOOP (log details).

Added a new dialogue manager unit test to test `validate_parse_tree` method. 

PR also contains some fixes for module paths, see #117. 

Fixes # (issue)
#145 
#117 (partial fix)

## Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [x] This is a unit test
- [ ] Documentation only change

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Example of action dictionary caught by validator when playing with bot (later fixed in ground truth):
P176155293

# Testing

In addition to conda setup, make sure modules are properly initialized with
`python setup.py develop`

Then run unit tests and test the bot as usual.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [x] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

**Comment by Rebecca-Qian** [comment]
FYI, removed `spacy` from `craftassist/requirements.txt` because it conflicts with base_agent dependencies. See P176017299 for error trace.

Doesn't seem to cause issues for me, cc @snyxan @kavyasrinet

**Comment by Rebecca-Qian** [comment]
Update: also fixed import paths in locobot tests, since some tests were failing.

cc @anuragprat1k

File: base_agent/nsp_dialogue_manager.py [hunk 1]
```diff
@@ -269,6 +294,16 @@ def get_logical_form(self, s: str, chat_as_list=False, ground_truth_actions: dic
             else:
                 d = self.model.parse(chat=s)
 
+        # Validate parse tree against grammar
+        is_valid_json = self.validate_parse_tree(d)
```
Comment (anuragprat1k) [inline base_agent/nsp_dialogue_manager.py:298]: how frequently do you expect this to be invalid? do we want to collect all the invalid parses - if so how?

File: craftassist/test/test_dialogue_manager.py [hunk 2]
```diff
@@ -80,6 +80,11 @@ def test_parses(self):
                 )
             )
 
+    def test_validate_json(self):
```
Comment (anuragprat1k) [inline craftassist/test/test_dialogue_manager.py:83]: add a successful validation test scenario as well.

File: craftassist/test/validate_json.py [hunk 3]
```diff
@@ -27,7 +27,16 @@ def validate_data(data_path, schema, resolver):
             except exceptions.ValidationError as e:
                 print(command)
                 pprint(parse_tree)
-                print("\n")                            
+                print("\n")
+
+
+def validate_instance(parse_tree, schema, resolver):
```
Comment (anuragprat1k) [inline craftassist/test/validate_json.py:33]: where is this used?

**Review by anuragprat1k — CHANGES_REQUESTED** [review]
The json validation changes look good. Requested some minor changes.

I don't see any of the `sys.path.append`s removed in this PR, so I don't understand why the tests were failing - did you figure out why?

**Review by kavyasrinet — COMMENTED** [review]
Thanks for making the change. Looks good. Let's add a unit test too.

File: base_agent/nsp_dialogue_manager.py (hunk 1 above)
Comment (Rebecca-Qian) [reply base_agent/nsp_dialogue_manager.py:298]: For ground truth, this should never be invalid, if it is, we need to fix ASAP. For parses from the semantic parser, it should not happen frequently. The model is generally good at producing parse trees that fit the grammar spec, even when they're wrong.

Agree, the wrong parses should be collected. I'll write to a separate file in datasets.

File: craftassist/test/validate_json.py (hunk 3 above)
Comment (Rebecca-Qian) [reply craftassist/test/validate_json.py]: Added docstring here

**Comment by Rebecca-Qian** [comment]
Updates:

- Added 2 test cases checking for spans with string values and arrays
- Added support for string value spans, since these are the format of spans in unit tests and currently autocomplete outputs. I think it's more readable to keep hand written parses in this format, so I decided to support different validation formats rather than enforce array spans everywhere.
- Refactored JSON validator and added docstrings

File: base_agent/nsp_dialogue_manager.py (hunk 1 above)
Comment (Rebecca-Qian) [reply base_agent/nsp_dialogue_manager.py:283]: Actually, may not need to write the failed parses to a separate file. Considering adding data checks to CI, then ground truth would always be up to date. Failed model parses aren't very actionable

**Review by anuragprat1k — APPROVED** [review]
looks good.

**Pull request merged**